### PR TITLE
adding OSR implementation

### PR DESCRIPTION
**Pull request by zhengshi1998**
The implementation of OSR algorithm has been added to the reporsitory. The detail of this algorithm can be found via https://dl.acm.org/doi/10.1145/3597503.3639099.

### CODE REVIEW

File: generateMetaInfo.py
Comment (umangm) [inline generateMetaInfo.py:1]: Is this file required? Does OSR not run on the same formats as other dynamic analysis algorithms in RAPID?

File: rapid.iml
Comment (umangm) [inline rapid.iml:1]: Remove and adjust the dependencies in `rapid.xml` instead

File: rapid1.iml
Comment (umangm) [inline rapid1.iml:1]: Remove and adjust the dependencies in `rapid.xml` instead

File: src/OSR.java
Comment (umangm) [inline src/OSR.java:1]: Can you please structure this file like other files (See for example `src/SyncPreserving.java`)?

File: src/engine/racedetectionengine/OSR/POBuild/NumArrayMin.java
Comment (umangm) [inline src/engine/racedetectionengine/OSR/POBuild/NumArrayMin.java:1]: Looks like `POBuild` is a standalone datastructure. If so, can it go to the `util` package of `rapid`?

File: src/engine/racedetectionengine/OSR/POBuild/NumArrayMin.java [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+						right = mid - 1;
+					} else {
+						return mid;
+					}
+				} else {
+					return 0;
+				}
+			}
+
+//			System.out.println("right : " + left + ", " + right + ", " + mid);
```
Comment (umangm) [inline src/engine/racedetectionengine/OSR/POBuild/NumArrayMin.java]:
```suggestion

```

File: src/engine/racedetectionengine/OSR/POBuild/NumArrayMin.java [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+						return mid;
+					}
+				} else {
+					return size - 1;
+				}
+			} else {
+				right = mid - 1;
+			}
+
+//			System.out.println("left : " + left + ", " + right + ", " + mid);
```
Comment (umangm) [inline src/engine/racedetectionengine/OSR/POBuild/NumArrayMin.java]:
```suggestion
```

File: src/engine/racedetectionengine/OSR/POBuild/NumArrayMin.java [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+		}
+
+		return left;
+	}
+
+	public int sumRange(int i, int j) {
+
+		if(this.nums.size() == 0) return -1;
+
+//		this.printArray();
```
Comment (umangm) [inline src/engine/racedetectionengine/OSR/POBuild/NumArrayMin.java]:
```suggestion
```

File: src/engine/racedetectionengine/OSR/POBuild/NumArrayMin.java [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+
+		int right = searchIndexLeft(i);
+		int left = searchIndexRight(j);
+
+//		System.out.println("input : " + i + ", " + j);
+//		System.out.println("query : " + left + ", " + right);
```
Comment (umangm) [inline src/engine/racedetectionengine/OSR/POBuild/NumArrayMin.java]:
```suggestion
```

File: src/engine/racedetectionengine/OSR/OSREvent.java [hunk 5]
```diff
@@ -0,0 +1,674 @@
...
+                if(i == j) continue;
+
+                Triplet<Integer, Integer, Long> testNode = nodes.get(j); // <thId, inThId, auxId>
+
+                if(testNode.first.intValue() == curNode.first.intValue()) {
+                    if (testNode.second > curNode.second) {
+                        graph.addEdge(curNode.third, testNode.third);
+                    }
+                } else if (vc.getClockIndex(testNode.first) != -1 && testNode.second >= vc.getClockIndex(testNode.first)) {
+//                    System.out.println("add edge : " + curNode.third + " -> " + testNode.third);
```
Comment (umangm) [inline src/engine/racedetectionengine/OSR/OSREvent.java]:
```suggestion
```

File: src/engine/racedetectionengine/OSR/OSREvent.java [hunk 6]
```diff
@@ -0,0 +1,674 @@
...
+            int lastRelThId = state.recentRelThreadId.get(lockId);
+
+            graph.addVertex(acqOAuxId);
+            nodes.add(new Triplet<>(backwardEdge.third, backwardEdge.second.inThreadId, acqOAuxId));
+            graph.addVertex(lastRelAuxId);
+            nodes.add(new Triplet<>(lastRelThId, lastRel.inThreadId, lastRelAuxId));
+            graph.addEdge(lastRelAuxId, acqOAuxId);
+        }
+
+//        System.out.println(nodes);
```
Comment (umangm) [inline src/engine/racedetectionengine/OSR/OSREvent.java]:
```suggestion
```

File: src/engine/racedetectionengine/OSR/OSREvent.java [hunk 7]
```diff
@@ -0,0 +1,674 @@
...
+//        if(!hasCycle) {
+//            System.out.println("graph size = " + state.graph.vertexSet().size() / 2);
+//        }
+
+        return !hasCycle;
+    }
+
+    public boolean buildGraph(OSRState state, List<Triplet<Integer, AcqEventInfo, Integer>> backwardEdges,
+                              AccessEventInfo e1, AccessEventInfo e2, int e1ThId, int e2ThId){
+        // backwardEdge = <lockId, acqO, thId of acqO>
```
Comment (umangm) [inline src/engine/racedetectionengine/OSR/OSREvent.java]:
```suggestion
```

File: src/engine/racedetectionengine/OSR/OSREvent.java [hunk 8]
```diff
@@ -0,0 +1,674 @@
...
+            return false;
+        }
+
+        List<Triplet<Integer, AcqEventInfo, Integer>> backwardEdges = getAllBackwardEdges(state);
+
+        boolean hasCycle = this.buildGraph(state, backwardEdges, e1, e2, e1ThId, e2ThId);
+
+//        if(!hasCycle) {
+//            System.out.println("graph size = " + state.graph.vertexSet().size() / 2);
+//        }
```
Comment (umangm) [inline src/engine/racedetectionengine/OSR/OSREvent.java]:
```suggestion
```

File: src/engine/racedetectionengine/OSR/OSREvent.java [hunk 9]
```diff
@@ -0,0 +1,674 @@
...
+
+                if(checkEventInVectorstamp(state.osrEventSet, e1, thId)){
+                    // e1 \in OSR set, should increase e1 till e1 \notin OSR
+                    continue;
+                }
+
+                if (checkRace(e1, e2, thId, this.getThread().getId(), state)) {
+//                    System.out.println("race between " + e1.auxId + ", " + e2.auxId);
+//                    updateRaceDistances(state, e1.auxId, e2.auxId);
+//                    System.out.println("OSR set between " + e1.auxId + ", " + e2.auxId + " " + state.osrEventSet);
```
Comment (umangm) [inline src/engine/racedetectionengine/OSR/OSREvent.java]:
```suggestion
```

File: src/engine/racedetectionengine/OSR/OSREvent.java [hunk 10]
```diff
@@ -0,0 +1,674 @@
...
```
Comment (umangm) [inline src/engine/racedetectionengine/OSR/OSREvent.java]:
```suggestion
```

File: src/OSR.java [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+        boolean time_reporting = true;
+        long startTimeAnalysis = 0;
+        if (time_reporting) {
+            startTimeAnalysis = System.currentTimeMillis(); // System.nanoTime();
+        }
+
+        POBuildReverse po = new POBuildReverse(trace_dir);
+        po.readTraceAndUpdateDS();
+        int numThreads = po.numThreads;
+//        po.printArray();
```
Comment (umangm) [inline src/OSR.java]:
```suggestion
```

File: src/rapid.iml
Comment (umangm) [inline src/rapid.iml:1]: This file can be deleted

File: src/util/NeibhorsHardCodeWordTricks.java
Comment (umangm) [inline src/util/NeibhorsHardCodeWordTricks.java:1]: Why was this deleted?

File: src/util/PairHardCodeWordTricks.java
Comment (umangm) [inline src/util/PairHardCodeWordTricks.java:1]: Why was this deleted?

**Review by umangm — CHANGES_REQUESTED** [review]
Please address my comments

File: generateMetaInfo.py
Comment (zhengshi1998) [reply generateMetaInfo.py:1]: This is requried, OSR needs some meta information to init data structures. This python file scans the trace once and produce the number of threads, locks, variables, total number of events and number of events per thread.

**Comment by zhengshi1998** [comment]
I have addressed these comments, please check now.

**Comment by umangm** [comment]
Optimistic Sync Reversal Races from ICSE’24

**Pull request merged**